wb_modbus_device_editor/template_manager.py

```diff
@@ -0,0 +1,189 @@
+                json_template, (""), path_mode="full", yajl_allow_comments=True
+            ):
+                dict_info = info[0]
+                basic_info = {
+                    "title": dict_info.get("title", None),
+                    "device_type": dict_info.get("device_type", None),
+                    "group": dict_info.get("device_type", None),
+                    "deprecated": dict_info.get("deprecated", None),
+                    "translates": dict_info.get("translates", None),
+                }
+                return basic_info
+
+    def _convert_list_to_dict(self, source):
+        if source == None or type(source) == dict:
+            return source
+
+        result = {}
+        for item in source:
+            id = item["id"]
+            item.pop("id")
+            result[id] = item
+        return result
+
+    def _get_template_full_info(self, template_path):
+        with open(template_path, encoding="utf-8") as json_template:
+            for info in jsonslicer.JsonSlicer(
+                json_template, (""), path_mode="full", yajl_allow_comments=True
+            ):
+                dict_info = info[0]
+                groups = dict_info["device"].get("groups")  # groups and parameters mey have dict type
+                parameters = dict_info["device"].get("parameters")
+                full_info = {
+                    "title": dict_info.get("title", None),
+                    "device": {
+                        "name": dict_info["device"]["name"],
+                        "groups": self._convert_list_to_dict(groups),
+                        "parameters": self._convert_list_to_dict(parameters),
+                        "translations": dict_info["device"].get("translations"),
+                        "setup": dict_info["device"].get("setup", None),
+                    },
+                }
+                return full_info
+
+    def update_properties(self):
+        self._properties = self._get_template_full_info(self._template_path)
+
+    def get_parameters_by_group_id(self, group_id) -> list:
+        if self._properties is None:
+            return None
+
+        parameters = self._properties["device"]["parameters"]
+        result = {}
+        for id, parameter in parameters.items():
+            if parameter.get("group") == group_id:
+                result.update({id: parameter})
+        return result
+
+    def get_parameter_enum(self, parameter_id) -> dict:
+        parameter = self._properties["device"]["parameters"][parameter_id]
+        enum = parameter["enum"]
+        enum_titles = parameter["enum_titles"]
+        for enum_title in enum_titles:
+            enum_title = self.translate(enum_title)
+
+        return {"enum": enum, "enum_titles": enum_titles}
+
+    def calc_parameter_condition(self, condition, values):
+
+        try:
+            condition = condition.replace("||", " or ").replace("&&", " and ").replace("!", " not ")
+            return eval(condition, {}, values)
```

Comment:
ууу, опасное. Прямо здесь и сейчас не очень понятно, чем именно это чревато (хотя кто-то может подсунуть некрасивый шаблон с подлянкой внутри), но лучше сразу перестраховаться и сделать хотя бы так:

```suggestion
            return eval(condition, {"__builtins__":None}, values)
```

https://lybniz2.sourceforge.net/safeeval.html

Это не абсолютно безопасно, но по крайней мере сильно ограничивает поверхность атаки

Reply:
исправлено